sql/schemachanger: model IndexColumn, avoid index backfill when unneeded

ajwerner opened the pull request:
**sql/schemachanger/rel: add support for Neq (!=)**

**sql/schemachanger: model IndexColumn explicitly**
This change normalizes the columns in an index. This lays the foundation to
handle cases where we add a NULL-able new column without a default value. In
these cases, we should just modify the existing index in-place. That work will
come in a follow-up commit. This will also prove valuable for more complex
add column scenarios.

**sql/schemachanger: add a special case for simple add column**
Before this change, when we added a column, we'd always create a new primary
index and swap into it. In the case where we're adding a nullable column with
no default value or computed expression, we have no reason to do this whole
dance. Indeed, backfilling that new index was a major regression for this
common case when compared to the legacy schema changer.

In this commit, we detect this special case and just add the new column to
the existing primary index.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83222)
<!-- Reviewable:end -->

ajwerner committed:
sql/schemachanger/rel: add support for Neq (!=)

This ends up being useful.

Release note: None

ajwerner committed:
sql/schemachanger: model IndexColumn explicitly

This change normalizes the columns in an index. This lays the foundation to
handle cases where we add a NULL-able new column without a default value. In
these cases, we should just modify the existing index in-place. That work will
come in a follow-up commit. This will also prove valuable for more complex
add column scenarios.

Release note: None

chengxiong-ruan reviewed (commented):
LGTM on everything outside of the `rel` which honestly I forgot how it works (I swear I knew it how it works 3 months ago).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-:-N5DXBhl7Hk1vM92YoOp:b-ffdacy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @chengxiong-ruan)

___
*[`pkg/sql/catalog/descpb/index_fetch.proto` line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5DK2SsC8ERpOda6EJK:-N5DK2SsC8ERpOda6EJL:b-gwmzs2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c412cdc9686e9d7caa742b795d5b624cd3cda16b/pkg/sql/catalog/descpb/index_fetch.proto#L47)):*
> ```protobuf
>   message KeyColumn {
>     optional Column column = 1 [(gogoproto.embed) = true, (gogoproto.nullable) = false];
>     optional catalog.catpb.IndexColumn.Direction direction = 2 [(gogoproto.nullable) = false];
> ```

I assume that there's no deserialization problem with this field type change because the essential type is actually the same and protobuf use field number to identify a value from disk?
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/index.go` line 346 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5DVWoI76eB4ms_Zivk:-N5DVWoI76eB4ms_Zivl:bsjgddb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c412cdc9686e9d7caa742b795d5b624cd3cda16b/pkg/sql/schemachanger/scexec/scmutationexec/index.go#L346)):*
> ```Go
> 		insertName(&id.StoreColumnNames)
> 	}
> 	// If this is a composite column, note that. Because we don't en
> ```

to be continued (sound like you wanted to say more heh)


<!-- Sent from Reviewable.io -->

ajwerner committed:
sql/schemachanger: add a special case for simple add column

Before this change, when we added a column, we'd always create a new primary
index and swap into it. In the case where we're adding a nullable column with
no default value or computed expression, we have no reason to do this whole
dance. Indeed, backfilling that new index was a major regression for this
common case when compared to the legacy schema changer.

In this commit, we detect this special case and just add the new column to
the existing primary index.

Release note: None

ajwerner commented:
That this last commit works in the rollback case was surprising to me given we don't explicitly remove the column with an operation from the primary index. It turns out that this gets covered by https://github.com/cockroachdb/cockroach/blob/cbfa501a444b67e553cd0d85f46f9288ca64dcd1/pkg/sql/schemachanger/scexec/scmutationexec/column.go#L197

I'm not sure I love this.

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go:
```diff
@@ -390,10 +419,33 @@ func addColumn(b BuildCtx, spec addColumnSpec) (backing *scpb.PrimaryIndex) {
 		b.Add(updatedPartitioning)
 	}
-
+	for _, ec := range append(existingColumns, ic) {
+		cloned := protoutil.Clone(ec).(*scpb.IndexColumn)
+		cloned.IndexID = temp.IndexID
+		b.Add(cloned)
+	}
```
Wow... Now we would have 3 sets of index columns for the old primary index, new primary index, and temp index; Or 5 sets if this newly added column is UNIQUE and hence another secondary index (and its temp index) needs to be created.

 The stage graph is going to get BIG!

Xiang-Gu commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go:
```diff
@@ -390,10 +419,33 @@ func addColumn(b BuildCtx, spec addColumnSpec) (backing *scpb.PrimaryIndex) {
+func getNextStoredIndexColumnOrdinal(
+	allTargets ElementResultSet, idx *scpb.PrimaryIndex,
+) (ord uint32) {
+	var foundAny bool
```
nit: just initialize return `ord = -1` we can get rid of `foundAny` and just `return ord ++`

Xiang-Gu commented on pkg/sql/schemachanger/scdecomp/decomp.go:
```diff
@@ -410,19 +410,37 @@ func (w *walkCtx) walkIndex(tbl catalog.TableDescriptor, idx catalog.Index) {
+				TableID:   tbl.GetID(),
+				IndexID:   idx.GetID(),
+				ColumnID:  c,
+				Ordinal:   uint32(i),
```
okay, the `ordinal` here is 0-indexing and different KIND of index columns have their `ordinal` relative in their own KIND

Xiang-Gu commented on pkg/sql/schemachanger/scexec/scmutationexec/index.go:
```diff
@@ -309,3 +293,74 @@ func (m *visitor) SetIndexName(ctx context.Context, op scop.SetIndexName) error
+	// order, but we'd need a way to express successor or something in rel
+	// rules to add those dependencies efficiently. Instead, we just don't
+	// and sort here.
+	id := index.IndexDesc()
```
nit: I think `idx` is a better name bc when I first read this block, my mind constantly mistakenly think it as some sort of ID

Xiang-Gu commented on pkg/sql/schemachanger/scexec/scmutationexec/index.go:
```diff
@@ -309,3 +293,74 @@ func (m *visitor) SetIndexName(ctx context.Context, op scop.SetIndexName) error
+	// and sort here.
+	id := index.IndexDesc()
+	n := int(op.Ordinal + 1)
+	insertName := func(s *[]string) {
```
nit: similarly, I suggest renaming those three lambda functions as `insertNameTo` bc I initially thought we are going to insert the input string array into somewhere else. Then I realize that we're inserting the col name/id/dir into the input string array.

Xiang-Gu commented on pkg/sql/schemachanger/scop/mutation.go:
```diff
@@ -541,3 +541,16 @@ type RefreshStats struct {
 }
+
+// AddColumnToIndex mutates an index to add a column to it.
+// The column should already exist on the table and so should
+// the index.
```
Side note: One confusion I start to get as I learned about the new schema changer is "what does it mean to say a column/index exists on a table". 

When I'm learning the old schema changer, I learned that if a table descriptor is undergoing a schema change, it will have mutations queued up in them. This means, when using the table descriptor to interpret the rows, we need to consult both the table indexes/columns as well as any queuing mutations. For example, in a table descriptor, if we can find a column 'k', but we also find a drop_column_k as its mutation that is write_and_delete_only, we say "column k does not exist as far as reading is concerned".

Now, with the new schema changer, we further introduce a place to state "status" in elements. For example, when a column is added, its associated elements are marked as "absent" and "toPublic". But I think a lot of operations in the new schema changer world will just treat this column as "already exists" and proceed to operate on their elements accordingly. 

Maybe the way I am approaching to understanding this is off and is actually a lot simpler. Can you try to clear it up a bit for me? What's your mental model (or, how do you think to yourself) when something (e.g. a table/index/column) "already exists" and when it "already no longer exists".

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_index_column.go:
```diff
@@ -0,0 +1,39 @@
+			scpb.Status_PUBLIC,
+			to(scpb.Status_ABSENT,
+				revertible(false)),
+		),
```
Why didn't we need something like `scop.RemoveColumnFromIndex` here when we need to drop a column? I guess we do need, and we will do it in `DROP COLUMN`?

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:
```diff
@@ -85,22 +85,65 @@ func init() {
 
 	registerDepRule(
-		"partitioning set right after temp index existence",
+		"partitioning and columns set right after temp index existence",
```
set aside IndexParitioning, how is this rule different from the above one where we state "temporary index must exist right before its dependents (really just its index columns) becomes public"?

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:
```diff
@@ -85,22 +85,65 @@ func init() {
+		func(from, to nodeVars) rel.Clauses {
+			return rel.Clauses{
+				from.el.Type(
+					(*scpb.PrimaryIndex)(nil),
+					(*scpb.SecondaryIndex)(nil),
```
The rule name says "*temp* index existence preceded index dependents" but why isn't the from type "scpb.TemporaryIndex"?

chengxiong-ruan replied:
Good catch :)

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:
```diff
@@ -85,22 +85,65 @@ func init() {
+					(*scpb.IndexColumn)(nil),
+				),
+				joinOnIndexID(from.el, to.el, "table-id", "index-id"),
+				targetStatus(from.target, scpb.Transient),
```
Shouldn't this be `scpb.Transient_Absent`?

chengxiong-ruan replied:
I think it is what you mean : http://github.com/cockroachdb/cockroach/blob/8601bedf62a7e0534a99c33087cdafcc6d1015fd/pkg/sql/schemachanger/scpb/constants.go#L41-L41

though good call that it is confusing.

Xiang-Gu reviewed (approved):
Thank you so much for this @ajwerner -- it will unblock a lot of PRs from me and generally make manipulating index and columns easier in the new schema changer.

I left some nit comments and questions. LGTM overall

postamar reviewed (commented):
I did a somewhat superficial pass and have only one substantial concern related to the drop-path of the new index-column element. 

> Reviewed 6 of 106 files at r1, 102 of 102 files at r2, 95 of 98 files at r3, 11 of 11 files at r4, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-:-N5Gp1vr4mh0-5yhZIAx:b1anl8e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @chengxiong-ruan, and @Xiang-Gu)

___
*[`pkg/sql/catalog/descpb/index_fetch.proto` line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5DK2SsC8ERpOda6EJK:-N5GjiPoBWlZww7Asxit:bp8eu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c412cdc9686e9d7caa742b795d5b624cd3cda16b/pkg/sql/catalog/descpb/index_fetch.proto#L47)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

I assume that there's no deserialization problem with this field type change because the essential type is actually the same and protobuf use field number to identify a value from disk?
</blockquote></details>

Yes, that's correct.
___
*[`pkg/sql/schemachanger/rel/query_lang.go` line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5GipAvEh20YI1szP8B:-N5GipAvEh20YI1szP8C:br49gk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba98086bd78ea99129ca454fbae6c80da6943054/pkg/sql/schemachanger/rel/query_lang.go#L44)):*
> ```Go
> //// attribute takes on a value of a different type than the passed value, a
> //// contradiction will be found; the entity's attribute must be bound, and it
> //// must be bound to a value of the same type as the provided value.
> ```

nit: s,////,//,
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 422 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5Tc-r4-422:-N5GlFAxBGLOq6ThC2M8:b-8q2m4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scdecomp/decomp.go#L422)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

okay, the `ordinal` here is 0-indexing and different KIND of index columns have their `ordinal` relative in their own KIND
</blockquote></details>

I was also a bit confused by this ordinal numbering scheme. Perhaps rename `Ordinal` to `OrdinalInKind` or something?
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_index_column.go` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5GkbWN3p5IINqXRgu7:-N5GkbWN3p5IINqXRgu8:b-vn8ugy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scplan/internal/opgen/opgen_index_column.go#L36)):*
> ```Go
> 			scpb.Status_PUBLIC,
> 			to(scpb.Status_ABSENT,
> 				revertible(false)),
> ```

I'm concerned that without a `notImplemented` op here we may be passing the rollback tests even when we shouldn't be. Also, this may hide some dep & op rules necessary for correct DROP behaviour.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5GmtYL7cUSHMe53JIr:-N5GmtYL7cUSHMe53JIs:b-y11q6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L97)):*
> ```Go
> 		})
> 	registerDepRule(
> 		"index existence precedes index dependents",
> ```

Adjust rule name. Perhaps "index exists right before index column"? Same comment applies elsewhere.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 520 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5GnG-a41LHSKCLXJ3F:-N5GnG-bD0AAXL02lBQq:b-qy6iaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L520)):*
> ```Go
> 					index, column, indexColumn, tableID, columnID, indexID,
> 				),
> 				index.AttrNeq(screl.SourceIndexID, catid.IndexID(0)),
> ```

Would it be worth introducing some kind of shorthand-rule for "ID is not set"? Just an idea. Perhaps premature.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
TFTR, RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-:-N5atKSy-qM9G7osHG6n:b-lhgni0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @fqazi, @postamar, and @Xiang-Gu)

___
*[`pkg/sql/catalog/descpb/index_fetch.proto` line 47 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5DK2SsC8ERpOda6EJK:-N5DYEcF79VJ7cysh0I6:b-b5ual0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c412cdc9686e9d7caa742b795d5b624cd3cda16b/pkg/sql/catalog/descpb/index_fetch.proto#L47)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

I assume that there's no deserialization problem with this field type change because the essential type is actually the same and protobuf use field number to identify a value from disk?
</blockquote></details>

Correct, should be the same. Protobufs when serialized are just field numbers. 
___
*[`pkg/sql/schemachanger/rel/query_lang.go` line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5GipAvEh20YI1szP8B:-N5HV08Y4VFj6sAURmXq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba98086bd78ea99129ca454fbae6c80da6943054/pkg/sql/schemachanger/rel/query_lang.go#L44)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: s,////,//,
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go` line 426 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5TW-r4-426:-N5GreRh5jYIzcqSJQHs:beqv7vx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go#L426)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

Wow... Now we would have 3 sets of index columns for the old primary index, new primary index, and temp index; Or 5 sets if this newly added column is UNIQUE and hence another secondary index (and its temp index) needs to be created.

 The stage graph is going to get BIG!
</blockquote></details>

Indeed. @fqazi is adding partition subzones, so even bigger! We're starting to out-grow our visualization technologies.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go` line 433 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5TW-r4-433:-N5GrurGDeuV8f-L3lJt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go#L433)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

nit: just initialize return `ord = -1` we can get rid of `foundAny` and just `return ord ++`
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go` line 357 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5TW-r5-357:-N5HV4GU4X20923r-UfC:b-ngjzrb) ([raw file](https://github.com/cockroachdb/cockroach/blob/45ef61357aae132c72b87e4781da9296612a142b/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go#L357)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

What if we added a column with ON-UPDATE expression and immediately followed by an `UPDATE` stmt in the same transaction? It's a scary realm where we have DDL stmts mixed up with DML stmts in a transaction that I never thought about before.
</blockquote></details>

it won't be populated because the column is not writable during the user transaction. I do understand that that may violate the user's expectations, but that doesn't change in this PR nor does it change with the declarative schema changer. It will get the default value. 
___
*[`pkg/sql/schemachanger/scdecomp/decomp.go` line 422 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5Tc-r4-422:-N5HcvwUDwvsDPzVdFzi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scdecomp/decomp.go#L422)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I was also a bit confused by this ordinal numbering scheme. Perhaps rename `Ordinal` to `OrdinalInKind` or something?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/index.go` line 346 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5DVWoI76eB4ms_Zivk:-N5DYKtI53kdVga1bg9b:b-lx4fj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c412cdc9686e9d7caa742b795d5b624cd3cda16b/pkg/sql/schemachanger/scexec/scmutationexec/index.go#L346)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

to be continued (sound like you wanted to say more heh)
</blockquote></details>

Fixed. 
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/index.go` line 315 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5Tk-r4-315:-N5Hcwsj37z3fsMqyhQN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scexec/scmutationexec/index.go#L315)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

nit: I think `idx` is a better name bc when I first read this block, my mind constantly mistakenly think it as some sort of ID
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/index.go` line 317 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5Tk-r4-317:-N5aavvG4uJG5JBZ7YDQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scexec/scmutationexec/index.go#L317)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

nit: similarly, I suggest renaming those three lambda functions as `insertNameTo` bc I initially thought we are going to insert the input string array into somewhere else. Then I realize that we're inserting the col name/id/dir into the input string array.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scop/mutation.go` line 547 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5Tl-r4-547:-N5GsC_YA6_d_F1Q5ejT:b-c5fc67) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scop/mutation.go#L547)):*

> "already exists" and when it "already no longer exists".

Here when I say exists in the key-value store somewhere beyond the element itself, generally as part of the descriptor, mutation or otherwise. Our mental models seem mostly aligned. 
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_index_column.go` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5GkbWN3p5IINqXRgu7:-N5GsMK46Dz63X79jdCm:bbw6ktj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scplan/internal/opgen/opgen_index_column.go#L36)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'm concerned that without a `notImplemented` op here we may be passing the rollback tests even when we shouldn't be. Also, this may hide some dep & op rules necessary for correct DROP behaviour.
</blockquote></details>

I'm instead going to implement this. The top level comment on the PR explains, to some extent, why I didn't need to do this. 
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_index_column.go` line 37 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5Tr-r4-37:-N5HWQyu9w1Ie0Kfu46l:beb0t6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scplan/internal/opgen/opgen_index_column.go#L37)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

Why didn't we need something like `scop.RemoveColumnFromIndex` here when we need to drop a column? I guess we do need, and we will do it in `DROP COLUMN`?
</blockquote></details>

https://github.com/cockroachdb/cockroach/blob/cbfa501a444b67e553cd0d85f46f9288ca64dcd1/pkg/sql/schemachanger/scexec/scmutationexec/column.go#L197 was the reason. I've now removed this. It required some fresh rules.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 97 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5GmtYL7cUSHMe53JIr:-N5aqir_AalxjOnwMs29:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L97)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Adjust rule name. Perhaps "index exists right before index column"? Same comment applies elsewhere.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 123 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5Tt-r4-123:-N5HWOQ18lJ_iN04SS2E:bdu5rt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L123)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Good catch :)
</blockquote></details>

copy-pasta, fixed
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 129 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5Tt-r4-129:-N5aay5lASV2EC2aCuU7:bae2qum) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L129)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

I think it is what you mean : http://github.com/cockroachdb/cockroach/blob/8601bedf62a7e0534a99c33087cdafcc6d1015fd/pkg/sql/schemachanger/scpb/constants.go#L41-L41

though good call that it is confusing.
</blockquote></details>

We can rename that value in a different PR if you desire
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 137 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5C3JiyEzLwITH-r5Tt-r4-137:-N5aqeYS2WA4sBIlJKJb:bwekrng) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L137)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

set aside IndexParitioning, how is this rule different from the above one where we state "temporary index must exist right before its dependents (really just its index columns) becomes public"?
</blockquote></details>

I've reworked the rules. Please do give them a read as they have changed.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 520 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/83222#-N5GnG-a41LHSKCLXJ3F:-N5ao-fU5Fh9Dyt_3vEe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9efa9fd1358da3818fe273773852438f6f8d892/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L520)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Would it be worth introducing some kind of shorthand-rule for "ID is not set"? Just an idea. Perhaps premature.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented:
TFTR!

bors r+

ajwerner commented:
bors r+

?
